core/lib/scmi-server/scmi_clock_consumer.c

```diff
@@ -0,0 +1,274 @@
+
+		res = clk_dt_get_by_index(fdt, subnode, 0, &clock);
+		if (res == TEE_ERROR_DEFER_DRIVER_INIT) {
+			panic("Unexpected init deferral");
+		} else if (res) {
+			EMSG("Can't get clock %s (%#"PRIx32"), skipped",
+			     fdt_get_name(fdt, subnode, NULL), res);
+			continue;
+		}
+
+		domain_id = fdt_reg_base_address(fdt, subnode);
+		s_clock = s_clocks + domain_id;
+		s_clock->domain_id = domain_id;
+
+		cuint = fdt_getprop(fdt, subnode, "domain-name", NULL);
+		if (cuint)
+			s_clock->domain_name = (char *)cuint;
+		else
+			s_clock->domain_name = fdt_get_name(fdt, subnode, NULL);
+
+		/* Check that the domain_id is not already used */
+		if (s_clock->clock) {
+			EMSG("Domain ID %"PRIu32" already used", domain_id);
+			panic();
+		}
+		s_clock->clock = clock;
+
+		/*
+		 * Get clock flags
+		 */
+		cuint = fdt_getprop(fdt, subnode, "flags", NULL);
+		if (cuint) {
+			s_clock->flags = fdt32_to_cpu(*cuint);
+			s_clock->flags &= SCMI_CLOCK_MASK;
+		}
+
+		DMSG("scmi clock shares %s on domain ID %"PRIu32,
+		     s_clock->domain_name, domain_id);
+	}
+
+	for (n = 0; n < s_clocks_count; n++) {
+		/*
+		 * Assign domain IDs to un-exposed clock as SCMI specification
+		 * requires the resource is defined even if not accessible.
+		 */
+		if (!s_clocks[n].clock) {
+			s_clocks[n].domain_id = n;
+			s_clocks[n].domain_name = "";
+		}
+
+		s_clocks[n].domain_name = strdup(s_clocks[n].domain_name);
+		if (!s_clocks[n].domain_name)
+			panic();
+	}
+
+	if (consumer->plat_clocks || channel_cfg->clock) {
+		EMSG("Clock already loaded: agent %u, channel %u",
+		     agent_cfg->agent_id, channel_cfg->channel_id);
+		panic();
+	}
+
+	consumer->plat_clocks_count = s_clocks_count;
+	consumer->plat_clocks = calloc(consumer->plat_clocks_count,
+				       sizeof(*consumer->plat_clocks));
+	assert(consumer->plat_clocks);
+
+	consumer->plat_clocks_flags =
+		calloc(consumer->plat_clocks_count,
+		       sizeof(*consumer->plat_clocks_flags));
+	assert(consumer->plat_clocks_flags);
+
+	channel_cfg->clock_count = s_clocks_count;
+	channel_cfg->clock = calloc(channel_cfg->clock_count,
+				    sizeof(*channel_cfg->clock));
+	assert(channel_cfg->clock);
+
+	for (i = 0; i < s_clocks_count; i++) {
+		unsigned int domain_id = s_clocks[i].domain_id;
+		struct clk *plat_clock = consumer->plat_clocks + domain_id;
+		uint32_t *plat_clock_flags = consumer->plat_clocks_flags +
+			domain_id;
+
+		*plat_clock_flags = s_clocks[i].flags;
+
+		plat_clock->ops = &plat_scmi_clk_ops;
+		plat_clock->priv = plat_clock_flags;
+		plat_clock->name = s_clocks[i].domain_name;
+		plat_clock->parent = s_clocks[i].clock;
+
+		if (*plat_clock_flags & SCMI_CLOCK_ALLOW_SET_RATE)
+			plat_clock->flags = CLK_SET_RATE_PARENT;
+
+		plat_clock->flags |= CLK_DUTY_CYCLE_PARENT;
+
+		if (clk_register(plat_clock))
+			panic();
+
+		channel_cfg->clock[domain_id] = (struct scmi_clock){
+			.name = clk_get_name(plat_clock),
+			.clk = plat_clock,
+			/* Default OFF for RIF issues */
+			.enabled = false,
```

Comment:
`.enabled = plat_clock_flags & SCMI_CLOCK_DEFAULT_ENABLED`
RIF issues should be handle by the RIF based platforms.